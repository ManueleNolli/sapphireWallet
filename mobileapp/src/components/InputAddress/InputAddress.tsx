import { Image, TouchableWithoutFeedback } from 'react-native'
import React, { useEffect } from 'react'
import { Input } from '@ui-kitten/components'
import { qrCodeSmall } from '../../assets/AssetsRegistry'
import { isAddress } from 'ethers'

type InputAddressProps = {
  value: string
  setValue: (value: string) => void
  isValid: boolean
  setIsValid: (isValid: boolean) => void
  setIsQRCodeScanning: (isQRCodeScanning: boolean) => void
}

export default function InputAddress({
  value,
  setValue,
  isValid,
  setIsValid,
  setIsQRCodeScanning,
}: InputAddressProps) {
  useEffect(() => {
    isAddress(value) ? setIsValid(true) : setIsValid(false)
  }, [value])

  const QrCodeIcon = (props: any): React.ReactElement => (
    <TouchableWithoutFeedback
      testID={'qr-code-icon'}
      onPress={() => {
        setIsQRCodeScanning(true)
      }}
    >
      <Image {...props} source={qrCodeSmall} />
    </TouchableWithoutFeedback>
  )

  const status = () => {
    if (isValid) {
      return 'success'
    } else if (value.length > 0) {
      return 'danger'
    }
    return 'basic'
  }

  return (
    <Input
      value={value}
      label="To"
      placeholder="Address"
      status={status()}
      accessoryRight={QrCodeIcon}
      onChangeText={(nextValue: string) => setValue(nextValue)}
    />
  )
}
