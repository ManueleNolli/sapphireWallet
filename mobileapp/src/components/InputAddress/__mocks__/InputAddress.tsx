const InputAddressMock = () => (
  <div data-testid={'InputAddress'}>InputAddress</div>
)

export default InputAddressMock
