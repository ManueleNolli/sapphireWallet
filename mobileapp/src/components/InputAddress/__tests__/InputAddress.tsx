import React from 'react'
import { act, fireEvent } from '@testing-library/react-native'
import InputAddress from '../InputAddress'
import renderWithTheme from '../../../TestHelper'
import { isAddress } from 'ethers'

// Mocking isAddress from 'ethers'
jest.mock('ethers')

describe('InputAddress Component', () => {
  it('renders correctly', () => {
    const tree = renderWithTheme(
      <InputAddress
        label="To"
        value=""
        setValue={() => {}}
        isValid={false}
        setIsValid={() => {}}
        setIsQRCodeScanning={() => {}}
      />
    )

    expect(tree.getAllByText('To')).toBeTruthy()
    expect(tree.getByPlaceholderText('Address')).toBeTruthy()
    expect(tree).toMatchSnapshot()
  })

  it('updates value and calls setIsValid correctly', async () => {
    ;(isAddress as unknown as jest.Mock).mockReturnValue(true)

    const setIsValidMock = jest.fn()
    const tree = renderWithTheme(
      <InputAddress
        label="to"
        value=""
        setValue={() => {}}
        isValid={false}
        setIsValid={setIsValidMock}
        setIsQRCodeScanning={() => {}}
      />
    )

    await act(async () => {
      fireEvent.changeText(tree.getByTestId('@undefined/input'), 'someAddress')
    })

    expect(isAddress).toHaveBeenCalled()
    expect(setIsValidMock).toHaveBeenCalledWith(true)
  })

  it('displays correct status', () => {
    const { getByTestId } = renderWithTheme(
      <InputAddress
        label=""
        value=""
        setValue={() => {}}
        isValid={false}
        setIsValid={() => {}}
        setIsQRCodeScanning={() => {}}
      />
    )

    expect(getByTestId('@undefined/input').props.status).toBe('basic')
  })

  it('displays success status when isValid is true', () => {
    ;(isAddress as unknown as jest.Mock).mockReturnValue(true)
    const { getByTestId } = renderWithTheme(
      <InputAddress
        label=""
        value="Address"
        setValue={() => {}}
        isValid
        setIsValid={() => {}}
        setIsQRCodeScanning={() => {}}
      />
    )

    expect(getByTestId('@undefined/input').props.status).toBe('success')
  })

  it('displays danger status when isValid is false and value length is greater than 0', () => {
    const { getByTestId } = renderWithTheme(
      <InputAddress
        label=""
        value="invalidAddress"
        setValue={() => {}}
        isValid={false}
        setIsValid={() => {}}
        setIsQRCodeScanning={() => {}}
      />
    )

    // Expecting the status to be 'danger'
    expect(getByTestId('@undefined/input').props.status).toBe('danger')
  })

  it('calls QrCodeIcon onPress correctly', () => {
    const setIsQRCodeScanningMock = jest.fn()
    const { getByTestId } = renderWithTheme(
      <InputAddress
        label=""
        value=""
        setValue={() => {}}
        isValid={false}
        setIsValid={() => {}}
        setIsQRCodeScanning={setIsQRCodeScanningMock}
      />
    )

    // Triggering QrCodeIcon onPress
    fireEvent.press(getByTestId('qr-code-icon'))
    expect(setIsQRCodeScanningMock).toHaveBeenCalledWith(true)
  })
})
