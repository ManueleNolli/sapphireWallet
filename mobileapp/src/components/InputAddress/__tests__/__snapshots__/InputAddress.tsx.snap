// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputAddress Component renders correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "opacity": 1,
    }
  }
  testID="@undefined/container"
>
  <Text
    appearance="default"
    category="p1"
    status="basic"
    style={
      [
        {
          "color": "#222B45",
          "fontFamily": "System",
          "fontSize": 15,
          "fontWeight": "400",
        },
        [
          {
            "color": "#8F9BB3",
            "fontFamily": "System",
            "fontSize": 12,
            "fontWeight": "800",
            "marginBottom": 4,
          },
          {
            "textAlign": "left",
          },
        ],
      ]
    }
  >
    To
  </Text>
  <View
    style={
      [
        {
          "backgroundColor": "#F7F9FC",
          "borderColor": "#E4E9F2",
          "borderRadius": 4,
          "borderWidth": 1,
          "minHeight": 40,
          "paddingHorizontal": 8,
          "paddingVertical": 7,
        },
        {
          "alignItems": "center",
          "flexDirection": "row",
          "width": "100%",
        },
      ]
    }
  >
    <TextInput
      appearance="default"
      editable={true}
      onBlur={[Function]}
      onChangeText={[Function]}
      onFocus={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      placeholder="Address"
      placeholderTextColor="#8F9BB3"
      size="medium"
      status="basic"
      style={
        [
          {
            "color": "#222B45",
            "fontFamily": "System",
            "fontSize": 15,
            "fontWeight": "normal",
            "marginHorizontal": 8,
          },
          {
            "flexBasis": "auto",
            "flexGrow": 1,
            "flexShrink": 1,
          },
          null,
          undefined,
        ]
      }
      testID="@undefined/input"
      value=""
    />
    <Image
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      source={1}
      style={
        {
          "height": 24,
          "marginHorizontal": 8,
          "tintColor": "#8F9BB3",
          "width": 24,
        }
      }
      testID="qr-code-icon"
    />
  </View>
</View>
`;
